libs/spock-shaded/src/main/java/org/codice/spock/extension/builtin/DeFinalizer.java

```diff
@@ -0,0 +1,189 @@
+ * Lesser General Public License as published by the Free Software Foundation, either version 3 of
+ * the License, or any later version.
+ *
+ * <p>This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details. A copy of the GNU Lesser General Public
+ * License is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+package org.codice.spock.extension.builtin;
+
+import java.lang.reflect.InvocationTargetException;
+import java.util.Arrays;
+import java.util.List;
+import org.codice.spock.extension.DeFinalizeWith;
+import org.junit.runner.Describable;
+import org.junit.runner.Description;
+import org.junit.runner.Runner;
+import org.junit.runner.manipulation.Filter;
+import org.junit.runner.manipulation.Filterable;
+import org.junit.runner.manipulation.NoTestsRemainException;
+import org.junit.runner.manipulation.Sortable;
+import org.junit.runner.manipulation.Sorter;
+import org.junit.runner.notification.RunNotifier;
+import org.junit.runners.JUnit4;
+import org.junit.runners.model.InitializationError;
+import org.spockframework.runtime.Sputnik;
+import spock.lang.Specification;
+
+/**
+ * The <code>Definalizer</code> test runner is designed as a generic proxy test runner for another
+ * JUnit test runner by indirectly instantiating that runner in order to add support for
+ * de-finalizing (i.e. removing the final constraint) 3rd party Java classes that needs to be mocked
+ * or stubbed during testing. It does so by creating a classloader designed with an aggressive
+ * strategy where it will load all classes first before delegating to its parent. This classloader
+ * will therefore reload all classes while definalizing those that are requested except for all
+ * classes in the following packages:
+ *
+ * <ul>
+ *   <li>java
+ *   <li>javax
+ *   <li>sun
+ *   <li>org.xml
+ *   <li>org.junit
+ * </ul>
+ *
+ * These packages are not being reloaded as they are required for this test runner to delegate to
+ * the real test runner. Even the actual test class will be reloaded in this internal classloader.
+ *
+ * <p>The indirect extension is done by means of delegation as the real test runner is instantiated
+ * from within the classloader that is created internally. This is to ensure that everything the
+ * test class, the test runner, and everything they indirectly reference are loaded from within the
+ * classloader.
+ *
+ * <p>This runner is especially useful with Spock where it is not possible to mock final methods as
+ * can be done with Mockito.
+ *
+ * <p>See <a href="https://github.com/spockframework/spock/issues/735"/>for a Spock enhancement
+ * request to support mocking final classes/methods at which point, this class will no longer be
```

Comment:
Should there be a Codice ticket to remove this if/when that feature is added?

Reply:
I would say let's open it when it does otherwise it will be a ticket that we will just end up closing anyway. 

That being said, this new version of the test runner is actually now usable on top of any other test runner and not just Spock. It did started as something useful for Spock but it can now be moved out and be reused even in standard JUnit.